pubsub/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py

```diff
@@ -565,34 +612,26 @@ def _on_response(self, response):
         ]
         self._dispatcher.modify_ack_deadline(items)
 
-        invoke_callbacks_for = []
+        with self._pause_resume_lock:
+            for received_message in response.received_messages:
+                message = google.cloud.pubsub_v1.subscriber.message.Message(
+                    received_message.message,
+                    received_message.ack_id,
+                    self._scheduler.queue,
+                )
+                self._messages_on_hold.put(message)
+                self._on_hold_bytes += message.size
 
-        for received_message in response.received_messages:
-            message = google.cloud.pubsub_v1.subscriber.message.Message(
-                received_message.message, received_message.ack_id, self._scheduler.queue
-            )
-            # Making a decision based on the load, and modifying the data that
-            # affects the load -> needs a lock, as that state can be modified
-            # by different threads.
-            with self._pause_resume_lock:
-                if self.load < _MAX_LOAD:
-                    invoke_callbacks_for.append(message)
-                else:
-                    self._messages_on_hold.put(message)
-                    self._on_hold_bytes += message.size
-
-            req = requests.LeaseRequest(ack_id=message.ack_id, byte_size=message.size)
-            self.leaser.add([req])
-            self.maybe_pause_consumer()
+                req = requests.LeaseRequest(
+                    ack_id=message.ack_id,
+                    byte_size=message.size,
+                    ordering_key=message.ordering_key,
+                )
+                self.leaser.add([req])
 
-        _LOGGER.debug(
-            "Scheduling callbacks for %s new messages, new total on hold %s (bytes %s).",
-            len(invoke_callbacks_for),
-            self._messages_on_hold.qsize(),
-            self._on_hold_bytes,
-        )
-        for msg in invoke_callbacks_for:
-            self._scheduler.schedule(self._callback, msg)
+            self._maybe_release_messages()
```

Comment:
Why do we maybe release messages when we actually receive new messages that increase the load?

Reply:
"Load" is the load from the user's perspective. So new messages that arrive and are put on hold do not increase the load.

We need to call _maybe_release_messages because we can't rely only on the user's completion of previous messages to release new messages. ie: the first time messages come in, we have to release them to the user.

Reply:
OK that makes sense, thanks.